src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs

```diff
@@ -236,9 +236,20 @@ private HttpMessageHandler SetupHandlerChain()
                 handler = new HttpProxyConnectionHandler(poolManager, settings._proxy, settings._defaultProxyCredentials, handler);
             }
 
-            if (settings._credentials != null || settings._allowAutoRedirect)
+            HttpMessageHandler unauthenticatedHandler = handler;
+
+            if (settings._credentials != null)
+            {
+                handler = new AuthenticationHandler(settings._preAuthenticate, settings._credentials, handler);
+            }
+
+            if (settings._allowAutoRedirect)
             {
-                handler = new AuthenticateAndRedirectHandler(settings._preAuthenticate, settings._credentials, settings._allowAutoRedirect, settings._maxAutomaticRedirections, handler);
+                // Just as with WinHttpHandler and CurlHandler, for security reasons, we do not support authentication on redirects
+                // if the credential is anything other than a CredentialCache.
+                // We allow credentials in a CredentialCache since they are specifically tied to URIs.
+                HttpMessageHandler redirectHandler = (settings._credentials is CredentialCache ? handler : unauthenticatedHandler);
+                handler = new RedirectHandler(settings._maxAutomaticRedirections, handler, redirectHandler);
```

Comment:
I don't understand this change.  If _credentials was set to a NetworkCredential, previously we would use those credentials for the first request, and then on redirect those credentials would not be used.  But it seems like with this change, we're never using those credentials, even for the initial request.  Am I misunderstanding?

Reply:
Oh, nevermind, I see you pass in both `handler` and `redirectHandler`.